### PR TITLE
span sampling rules

### PR DESCRIPTION
**Pull request by dgoffredo**
These changes implement this [internal RFC][1] that describes a replacement for "app analytics."

[1]: https://github.com/DataDog/architecture/blob/master/rfcs/apm/integrations/single-span-ingestion-control/README.md

### CODE REVIEW

**Comment by dgoffredo** [comment]
Caleb: Fuzz the glob function.

File: test/tracer_factory_test.cpp [hunk 1]
```diff
@@ -3,12 +3,82 @@
...
 
 #include "../src/tracer.h"
 #include "mocks.h"
 
 using namespace datadog::opentracing;
 
+namespace {
+
+// `EnvGuard` sets an environment variable to a specified value for the scope
+// of the `EnvGuard`, restoring the environment variables previous value
```
Comment (dgoffredo) [inline test/tracer_factory_test.cpp:17]: variable's

File: test/tracer_factory_test.cpp [hunk 2]
```diff
@@ -3,12 +3,82 @@
...
+// the file at the end of the `TmpFile` scope.  A path to the temporary file is
+// accesible via the `name` member function.
+class TmpFile {
+  std::FILE *file_;
+  std::string name_;
+
+ public:
+  explicit TmpFile(ot::string_view content) {
+    char name_buffer[L_tmpnam];
+    if (std::tmpnam(name_buffer) == nullptr) {
```
Comment (dgoffredo) [inline test/tracer_factory_test.cpp:57]: This emits a warning during the build; I don't know how to suppress it.

The alternative is to use a POSIX function outside of the C++ standard.  The advantage for keeping this in the language standard is that it won't cause issues for Windows builds.

File: doc/configuration.md [hunk 3]
```diff
@@ -253,6 +253,30 @@ made, is propagated between services along the trace in the form of the
 order to prevent rejection by peers or other HTTP header policies.  This
 configuration option is that limit, in bytes.
 
+### Span Sampling Rules
+When a trace is dropped, it may still be advantageous to send some of its spans
```
Comment (cgilmour) [inline doc/configuration.md:257]: Consider restructuring this without the term "advantageous".
A simpler explanation of what it does, examples of cases where it's the right thing to do and link to how to do so should be enough.

File: doc/sampling.md [hunk 4]
```diff
@@ -120,5 +120,37 @@ This configuration option has the same meaning as the `DD_TRACE_RATE_LIMIT`
...
+    "max_per_second": <limit in spans sampled by this rule each second, or unlimited if absent>
+}, ...]
+```
+
+The `service` and `name` are glob patterns, where "glob" here means:
+- `*` matches any substring, including the empty string,
+- `?` matches exactly one of any character, and
+- any other character matches exactly one of itself.
+
+Span sampling rules are examined only when the enclosing trace is to be
```
Comment (cgilmour) [inline doc/sampling.md:144]: Does this need to say "automatically dropped" or is it still applied for manual drop?

File: doc/sampling.md [hunk 5]
```diff
@@ -120,5 +120,37 @@ This configuration option has the same meaning as the `DD_TRACE_RATE_LIMIT`
...
+dropped.
+
+The first span sampling rule that matches a span is used to make a span
+sampling decision for that span.  If the decision is "keep," then the span is
+sent to Datadog despite the enclosing trace having been dropped.
+
+Span sampling rules can be configured [directly][3] or [in a file][4].
```
Comment (cgilmour) [inline doc/sampling.md:151]: Are there reference examples for these somewhere?
A representative example of what we'd want users to do (and why) is a helpful starting point.

**Review by cgilmour — APPROVED** [review]
A handful of things to consider, but overall approved. Thanks!
If some changes trigger github's need to re-approve, let me know.

File: src/sampling_mechanism.h [hunk 6]
```diff
@@ -59,7 +59,12 @@ enum class SamplingMechanism {
   // Reserved for future use.
   AppSec = 5,
   // Reserved for future use.
-  RemoteRateUserDefined = 6
+  RemoteRateUserDefined = 6,
+  // Reserved for future use.
+  RemoteRateEmergency = 7,
+  // Individual span kept by a matching span sampling rule when the enclosing
+  // trace was dropped.
+  SpanRule = 8
```
[the remark this answers is not in the document]
Comment (dgoffredo) [reply src/sampling_mechanism.h:67]: Apparently yes (TIL): https://stackoverflow.com/questions/6372650/trailing-commas-and-c

File: src/tracer.cpp [hunk 7]
```diff
@@ -223,16 +224,19 @@ Tracer::Tracer(TracerOptions options, std::shared_ptr<SpanBuffer> buffer, TimePr
...
     : logger_(logger),
       opts_(options),
       get_time_(getRealTime),
       get_id_(getId),
       legacy_obfuscation_(legacyObfuscationEnabled()) {
-  configureRulesSampler(sampler);
+  assert(logger_);
+  configureRulesSampler(trace_sampler);
+  auto span_sampler = std::make_shared<SpanSampler>();
+  span_sampler->configure(opts_.span_sampling_rules, *logger_, get_time_);
```
[the remark this answers is not in the document]
Comment (dgoffredo) [reply src/tracer.cpp:236]: Which raw pointer?

File: doc/sampling.md (hunk 4 above)
Comment (dgoffredo) [reply doc/sampling.md:144]: Span sampling rules apply when the trace is dropped by the tracer for any reason, including a manual drop.

Seems odd, right?  I said drop this, but I also said to sometimes keep part of it when we drop it.

Now I'm thinking "well, it's only the statistical trace sampling decisions that customers want to counter with single span sampling, so let's exclude manually dropped traces."  But then I think "ah, but we no longer consider rule-based (i.e. statistics set by the user, e.g. DD_TRACE_SAMPLE_RATE) sampling decisions 'auto', now they're 'user'."

File: doc/sampling.md (hunk 5 above)
Comment (dgoffredo) [reply doc/sampling.md:151]: Will do

File: doc/sampling.md (hunk 4 above)
Comment (dgoffredo) [reply doc/sampling.md:144]: I suppose that once we have sampling mechanism propagation (as we recently merged but have not yet released), we could distinguish between "user said drop" and "user said less than 100% and we're being dropped."  Even then, though, we couldn't count on a service sending us a sampling mechanism alongside the sampling priority.

File: src/glob.cpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
+#include "glob.h"
+
+#include <cstdint>
+
+namespace datadog {
+namespace opentracing {
+
+bool glob_match(ot::string_view pattern, ot::string_view subject) {
```
[the remark this answers is not in the document]
Comment (dgoffredo) [reply src/glob.cpp:8]: Worth it also for the sake of dipping my toe in the libfuzzer pond.

File: include/datadog/opentracing.h [hunk 9]
```diff
@@ -156,6 +156,47 @@ struct TracerOptions {
...
+  //   order for the rule to match.  If "name" is not specified, then its default
+  //   value is "*".  Glob patterns are described below.
+  // - "sample_rate" is the probability that a span matching the rule will be
+  //   kept.  If "sample_rate" is not specified, then its default value is 1.0.
+  // - "max_per_second" is the maximum number of spans that will be kept on
+  //   account of this rule each second.  Spans that would cause the limit to
+  //   be exceeded are dropped.  If "max_per_second" is not specified, then
+  //   there is no limit.
+  //
+  // Glob patterns are strings that are evaluated character-by-character
```
[the remark this answers is not in the document]
Comment (dgoffredo) [reply include/datadog/opentracing.h]: After a cursory googling, I didn't find a description whose wording was as direct as we would like, and then there's the risk of tricking the user into thinking it's some other language (e.g. PCRE, BRE, ERE,  POSIX shell Pattern Matching Notation...).

Next best thing, I referred to it as a "simplified form of regular expressions" and then described the wildcard characters.

I went through many iterations of describing it in terms of sets, grammars, etc.

This is better.

**Pull request merged**